etherman/etherman.go

```diff
@@ -1428,12 +1431,74 @@ func decodeSequencedBatches(smcAbi abi.ABI, txData []byte, forkID uint64, lastBa
 			}
 			sequencedBatches[i] = batch
 		}
+
 		return sequencedBatches, nil
 	}
 
 	return nil, fmt.Errorf("unexpected method called in sequence batches transaction: %s", method.RawName)
 }
 
+type batchInfo struct {
+	num      uint64
+	hash     common.Hash
+	isForced bool
+}
+
+func retrieveBatchData(da dataavailability.BatchDataProvider, batchInfos []batchInfo, daMessage []byte) ([][]byte, error) {
+	validiumData, err := getBatchData(da, batchInfos, false, daMessage)
+	if err != nil {
+		return nil, err
+	}
+	forcedData, err := getBatchData(da, batchInfos, true, nil)
+	if err != nil {
+		return nil, err
+	}
+	data := make([][]byte, len(batchInfos))
+	for i, info := range batchInfos {
+		bn := info.num
+		if info.isForced {
+			data[i] = forcedData[bn] // these are guaranteed to be in map
+		} else {
+			data[i] = validiumData[bn] // these are guaranteed to be in map
+		}
+	}
+	return data, nil
+}
+
+// getBatchData picks either validium or forced batch data (based on forced arg) and returns a map of batch number -> data, or error
+func getBatchData(da dataavailability.BatchDataProvider, batchInfos []batchInfo, forced bool, daMessage []byte) (map[uint64][]byte, error) {
+	var batchNums []uint64
+	var batchHashes []common.Hash
+	for _, info := range batchInfos {
+		if info.isForced == forced {
+			batchNums = append(batchNums, info.num)
+			batchHashes = append(batchHashes, info.hash)
+		}
+	}
+	data := make(map[uint64][]byte)
+	if len(batchNums) == 0 {
+		return data, nil
+	}
+	var batchL2Data [][]byte
+	var err error
+	if forced {
+		batchL2Data, err = da.GetForcedBatchL2Data(batchNums, batchHashes)
```

Comment:
This should be queried directly to the stateDB, as it's guaranteed that you will have this data on your local DB, since the Etherman is already coupled with the StateDB, it feels a bit unnecessary to go through the DA. And for sure you shouldn't ask for this data to a remote node, but just fetch it from your local DB

Reply:
Ah, your point that the forced stuff should not come from that data availability interface.